tests/integration/test_shared_db.py

```diff
@@ -0,0 +1,253 @@
+        users_that_should_not_exist: List of users that should not exist in the database
+    """
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    select_users_sql = [
+        "SELECT CONCAT(user, '@', host) FROM mysql.user",
+    ]
+
+    unit = ops_test.model.applications[APP_NAME].units[0]
+    unit_address = await unit.get_public_address()
+
+    # Retrieve all users in the database
+    output = await execute_commands_on_unit(
+        unit_address,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+        select_users_sql,
+    )
+
+    # Assert users that should exist
+    for user in users_that_should_exist:
+        assert user in output
+
+    # Assert users that should not exist
+    for user in users_that_should_not_exist:
+        assert user not in output
+
+
+@pytest.mark.order(100)
+@pytest.mark.abort_on_fail
+async def test_keystone_bundle(ops_test: OpsTest) -> None:
+    """Deploy the keystone bundle to test the 'shared-db' relation.
+
+    Args:
+        ops_test: The ops test framework
+    """
+    # Build and deploy the mysql charm
+    charm = await ops_test.build_charm(".")
+    config = {"cluster-name": CLUSTER_NAME}
+    await ops_test.model.deploy(charm, application_name=APP_NAME, config=config, num_units=3)
+
+    # Reduce the update_status frequency until the keystone charm is deployed
+    async with ops_test.fast_forward():
+
+        # Wait until the mysql charm is successfully deployed
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[APP_NAME].units) == 3
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=True,
+            timeout=1000,
+        )
+        assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+        for unit in ops_test.model.applications[APP_NAME].units:
+            assert unit.workload_status == "active"
+
+        # Get the server config credentials
+        random_unit = ops_test.model.applications[APP_NAME].units[0]
+        server_config_credentials = await get_server_config_credentials(random_unit)
+
+        # Deploy and test the first deployment of keystone
+        await deploy_and_relate_keystone_with_mysql(ops_test, KEYSTONE_APP_NAME, 2)
+        await check_successful_keystone_migration(ops_test, server_config_credentials)
+
+        keystone_users = []
+        for unit in ops_test.model.applications[KEYSTONE_APP_NAME].units:
+            unit_address = await unit.get_public_address()
+
+            keystone_users.append(f"keystone@{unit_address}")
+
+        await check_keystone_users_existence(
+            ops_test, server_config_credentials, keystone_users, []
+        )
+
+        # Deploy and test another deployment of keystone
```

Comment:
This test is pretty big. It might be useful to split it into a few tests, so we can easily pinpoint what functionality causes it to fail.

Reply:
Indeed is huge - will evaluate what I can do about it